### PR TITLE
Add shortened values to economy system

### LINKED ISSUE #5352: Support shortened numbers in /pay, /bal and /baltop (1k = 1000, 3.44m = 3440000, etc)
### Feature description

Many servers with economies have players dealing with high numbers, even beyond trillions.

It would be a nice QoL feature if the /pay command supported shortened number format.

Instead of /pay iLemon 115000000000000
I could type /pay iLemon 115t

Support K, M, B, T, Q (case insensitive)

Also, have a config option so that the numbers in /bal and /baltop are shortened too.

### How the feature is useful

It saves players from typing long numbers, especially on servers with economies that go into trillions and quadrillions.

### PR DESCRIPTION
**Pull request by seppe-vandenberk**
### Information

This PR closes #5352. 

### Details

**Proposed feature:**  

Adding shortened values for the /pay command like 115m instead of 115000000, all values upto Q (Quaddrillion are covered). Also /bal and /baltop show shortened numbers when the option is true in the config.yml. 

**Environments tested:**    

OS: Microsoft Windows 11

Java version: java version "18.0.2" 2022-07-19  

- [x] Most recent Paper version (1.XX.Y, git-Paper-BUILD)
- [ ] CraftBukkit/Spigot/Paper 1.12.2
- [ ] CraftBukkit 1.8.8

**Demonstration:**  

image
image

### CODE REVIEW

**Comment by Atticuss26** [comment]
I would like this

File: Essentials/src/main/java/com/earth2me/essentials/Settings.java [hunk 1]
```diff
@@ -1943,4 +1943,10 @@ public boolean isUpdateCheckEnabled() {
     public boolean showZeroBaltop() {
         return config.getBoolean("show-zero-baltop", true);
     }
+
+    @Override
+    public boolean isShortNumbersAllowed(){
+        return config.getBoolean("shortened-economy-numbers",true);
```
Comment (powercasgamer) [inline Essentials/src/main/java/com/earth2me/essentials/Settings.java:1949]:
```suggestion
        return config.getBoolean("shortened-economy-numbers", true);
```

File: .idea/checkstyle-idea.xml
Comment (powercasgamer) [inline .idea/checkstyle-idea.xml:1]: Shouldn't be deleted

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandpay.java [hunk 2]
```diff
@@ -33,12 +32,49 @@ public void run(final Server server, final User user, final String commandLabel,
         }
 
         final String stringAmount = args[1].replaceAll("[^0-9\\.]", "");
+        final String decimalFactor;
 
         if (stringAmount.length() < 1) {
             throw new NotEnoughArgumentsException();
         }
+        if(ess.getSettings().isShortNumbersAllowed()){
```
Comment (powercasgamer) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandpay.java:40]:
```suggestion
        if (ess.getSettings().isShortNumbersAllowed()) {
```

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandpay.java [hunk 3]
```diff
@@ -33,12 +32,49 @@ public void run(final Server server, final User user, final String commandLabel,
...
+                    break;
+                case "q":
+                case "Q":
+                    decimalFactor = "1 000 000 000 000 000".replaceAll(" ","");
+                    break;
+                default:
+                    decimalFactor = "1";
+                    break;
+            }
+        } else{
```
Comment (powercasgamer) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandpay.java:72]:
```suggestion
        } else {
```

File: Essentials/src/main/java/com/earth2me/essentials/utils/NumberUtil.java [hunk 4]
```diff
@@ -66,9 +66,17 @@ public static String formatAsPrettyCurrency(final BigDecimal value) {
     }
 
     public static String displayCurrency(final BigDecimal value, final IEssentials ess) {
-        String currency = formatAsPrettyCurrency(value);
+        String factor = "";
+        BigDecimal money = value;
+
+        if(ess.getSettings().isShortNumbersAllowed()) {
```
Comment (powercasgamer) [inline Essentials/src/main/java/com/earth2me/essentials/utils/NumberUtil.java:72]:
```suggestion
        if (ess.getSettings().isShortNumbersAllowed()) {
```

File: Essentials/src/main/java/com/earth2me/essentials/utils/NumberUtil.java [hunk 5]
```diff
@@ -135,4 +151,57 @@ public static boolean isNumeric(final String sNum) {
...
+            factor = "M";
+        } else if (value.compareTo(decimalThousand) > 0) {
+            factor = "K";
+        }
+
+        return factor;
+    }
+
+    public static BigDecimal shortenNumbers(BigDecimal value, String factor) {
+
```
Comment (powercasgamer) [inline Essentials/src/main/java/com/earth2me/essentials/utils/NumberUtil.java:180]:
```suggestion
```

File: Essentials/src/main/java/com/earth2me/essentials/utils/NumberUtil.java [hunk 6]
```diff
@@ -79,9 +87,17 @@ public static String displayCurrency(final BigDecimal value, final IEssentials e
     }
 
     public static String displayCurrencyExactly(final BigDecimal value, final IEssentials ess) {
-        String currency = value.toPlainString();
+        String factor = "";
+        BigDecimal money = value;
+
+        if(ess.getSettings().isShortNumbersAllowed()) {
```
Comment (powercasgamer) [inline Essentials/src/main/java/com/earth2me/essentials/utils/NumberUtil.java:93]:
```suggestion
        if (ess.getSettings().isShortNumbersAllowed()) {
```

**Review by powercasgamer — CHANGES_REQUESTED** [review]

**Comment by Evidentsinger14** [comment]
closing in favor of #5457

**Pull request closed without merging**